soccer/gameplay/skills/dribble.py

```diff
@@ -0,0 +1,126 @@
+import single_robot_composite_behavior
+import behavior
+import robocup
+import role_assignment
+import main
+import constants
+import time
+import enum
+import skills.aim
+import skills.capture
+
+
+## Behavior that moves the ball to a specified location
+class Dribble(single_robot_composite_behavior.SingleRobotCompositeBehavior):
+    class State(enum.Enum):
+        capture = 1
+        aim = 2
+        drive = 3
+
+    def __init__(self, pos=None, vel=0):
+        super().__init__(continuous=False)
+
+        for state in Dribble.State:
+            self.add_state(state, behavior.Behavior.State.running)
+
+        self._threshold = 0.1  #default value matches the required accuracy for a placement command
+        self._pos = pos
+        self._vel = vel
+
+        self.add_transition(behavior.Behavior.State.start,
+                            Dribble.State.capture, lambda: True, 'immediately')
+
+        self.add_transition(
+            Dribble.State.capture, Dribble.State.aim,
+            lambda: self.subbehavior_with_name('capture').state == behavior.Behavior.State.completed,
+            'done capturing')
+
+        self.add_transition(
+            Dribble.State.capture, behavior.Behavior.State.completed,
+            lambda: (main.ball().pos - self._pos).mag() < self._threshold and main.ball().vel.mag() < .1,
+            'ball is already in target')
+
+        #Put the ball between the robot and the target
+        self.add_transition(Dribble.State.aim, Dribble.State.drive,
+                            lambda: self.aimed() and self.robot.has_ball(),
+                            'done aiming')
+
+        self.add_transition(Dribble.State.aim, Dribble.State.capture,
+                            lambda: self.fumbled(), 'fumbled')
+
+        self.add_transition(Dribble.State.drive, Dribble.State.capture,
+                            lambda: self.fumbled(), 'fumbled')
+
+        self.add_transition(
+            Dribble.State.drive, behavior.Behavior.State.completed,
+            lambda: (main.ball().pos - self._pos).mag() < self._threshold and not self.fumbled() and main.ball().vel.mag() < .05,
+            'finished driving')
+
+        self.last_ball_time = 0
+
+    def fumbled(self):
+        return not self.robot.has_ball() and time.time(
+        ) - self.last_ball_time > 0.3
+
+    ## the position to move to (a robocup.Point object)
+    @property
+    def pos(self):
+        return self._pos
+
+    @property
+    def vel(self):
+        return self._vel
+
+    @pos.setter
+    def pos(self, value):
+        self._pos = value
+
+    @vel.setter
+    def vel(self, value):
```

Comment:
Jay wanted to make this skill to be usable outside of just a placement so I had set it up to allow you to define an end velocity but when  I removed move_direct_end_vel it became useless so I will go ahead and take it out.